akka-http-core/src/main/scala/akka/http/impl/engine/ws/WebSocketClientBlueprint.scala

```diff
@@ -57,68 +57,82 @@ object WebSocketClientBlueprint {
+
+      override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+        new GraphStageLogic(shape) with InHandler with OutHandler {
+
+          def initial: InHandler = parsingResponse
+
+          def parsingResponse: InHandler = new InHandler {
+            // a special version of the parser which only parses one message and then reports the remaining data
+            // if some is available
+            val parser = new HttpResponseParser(settings.parserSettings, HttpHeaderParser(settings.parserSettings)()) {
+              var first = true
+              override def handleInformationalResponses = false
+              override protected def parseMessage(input: ByteString, offset: Int): Unit = {
+                if (first) {
+                  first = false
+                  super.parseMessage(input, offset)
+                } else {
+                  emit(RemainingBytes(input.drop(offset)))
+                  terminate()
+                }
+              }
             }
-          }
-        }
-        parser.setContextForNextResponse(HttpResponseParser.ResponseContext(HttpMethods.GET, None))
-
-        def onPush(elem: ByteString, ctx: Context[ByteString]): SyncDirective = {
-          parser.parseBytes(elem) match {
-            case NeedMoreData ⇒ ctx.pull()
-            case ResponseStart(status, protocol, headers, entity, close) ⇒
-              val response = HttpResponse(status, headers, protocol = protocol)
-              Handshake.Client.validateResponse(response, subprotocol.toList, key) match {
-                case Right(NegotiatedWebSocketSettings(protocol)) ⇒
-                  result.success(ValidUpgrade(response, protocol))
-
-                  become(transparent)
-                  valve.open()
-
-                  val parseResult = parser.onPull()
-                  require(parseResult == ParserOutput.MessageEnd, s"parseResult should be MessageEnd but was $parseResult")
-                  parser.onPull() match {
-                    case NeedMoreData          ⇒ ctx.pull()
-                    case RemainingBytes(bytes) ⇒ ctx.push(bytes)
-                    case other ⇒
-                      throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
+            parser.setContextForNextResponse(HttpResponseParser.ResponseContext(HttpMethods.GET, None))
+
+            def onPush(): Unit = {
+              val elem = grab(in)
+              parser.parseBytes(elem) match {
+                case NeedMoreData ⇒ pull(in)
+                case ResponseStart(status, protocol, headers, entity, close) ⇒
+                  val response = HttpResponse(status, headers, protocol = protocol)
+                  Handshake.Client.validateResponse(response, subprotocol.toList, key) match {
+                    case Right(NegotiatedWebSocketSettings(protocol)) ⇒
+                      result.success(ValidUpgrade(response, protocol))
+
+                      become(transparent)
```

Comment:
as you now know, become is gone (please compile the code before submitting PRs :-)), so we'll need to replace it with `setHandler(in, transparent)` :)